fix(ivy): content projection with Shadow DOM not working

crisbeto opened the pull request:
Fixes components with native content projection (using `<content>` or `<slot>`) not working under Ivy.

The issue comes from the fact that when creating elements inside a component, we sometimes don't append the element immediately, but we leave it to projection to move it into its final destination. This ends up breaking the native projection, because the slots have to be in place from the beginning. The following changes switch to appending the element immediately when inside a component with Shadow DOM encapsulation.

This PR resolves FW-841.

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -511,15 +514,24 @@ function getRenderParent(tNode: TNode, currentView: LView): RElement|null {
-    } else {
-      return getNativeByTNode(parent, currentView) as RElement;
+      const componentInstance = getComponentAtNodeIndex(parent.index, currentView);
+      const componentDef = componentInstance && getComponentDef(componentInstance.constructor);
```
`getComponentDef` has a megamorphic property access inside it, so let's avoid using it. There is another way to get a component def for a node:

```ts
const tData = currentView[TVIEW].data;
const tNode = tData[parent.index] as TNode;
const encapsulation = tData[tNode.directiveStart].encapsulation;
```

kara reviewed (approved):
LGTM

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -6,9 +6,12 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {ViewEncapsulation} from '../core';
```
This import is causing a "cyclical dependency" error between files in G3. Can you drill down to the `metadata/view` file?

alxhub commented:
Needs a rebase :) thanks!

crisbeto committed:
fix(ivy): content projection with Shadow DOM not working

Fixes components with native content projection (using `<content>` or `<slot>`) not working under Ivy.

The issue comes from the fact that when creating elements inside a component, we sometimes don't append the element immediately, but we leave it to projection to move it into its final destination. This ends up breaking the native projection, because the slots have to be in place from the beginning. The following changes switch to appending the element immediately when inside a component with Shadow DOM encapsulation.

This PR resolves FW-841.

crisbeto commented:
Rebased.

kara commented:
merge assistance: looks like the merge status is stuck, though all checks pass.
